### ISSUE TITLE
Add a WeakSender type

### ISSUE THREAD

**Issue by canndrew** [rust-lang/rfcs#1549]
The `mpsc` module should have a sender type that holds a weak reference to the channel. A `WeakSender` is to a `Sender` what a `sync::Weak` is to a `sync::Arc`. The channel will close if all `Sender`s are dropped even if the `Receiver` is still alive.

The API for this could look similar to the `Arc` API. Just add a method `Sender::downgrade(self) -> WeakSender<T>`. This should be easy to implement given that `Sender` currently uses an `Arc` internally to hold it's reference to the channel.

**Comment by AaronCAlbers** [comment]
I agree that there should be a ```WeakSender<T>``` and a ```WeakSyncSender<T>```. That way we can break circular references.

This would be especially useful in a Flow Based Programming implementation with components that connect there outputs directly to there inputs (or through some longer cycle). These cycles are very easy to detect and could easily be removed with this feature.
